sql/src/main/java/io/crate/execution/engine/collect/GroupByOptimizedIterator.java

```diff
@@ -0,0 +1,357 @@
+                        } else {
+                            aggregateValues(aggregations, ramAccounting, states);
+                        }
+                        if (values.nextOrd() != SortedSetDocValues.NO_MORE_ORDS) {
+                            throw new GroupByOnArrayUnsupportedException(keyIndexFieldData.getFieldName());
+                        }
+                    } else {
+                        if (nullStates == null) {
+                            nullStates = initStates(bigArrays, aggregations, ramAccounting);
+                        } else {
+                            aggregateValues(aggregations, ramAccounting, nullStates);
+                        }
+                    }
+                }
+                for (long ord = 0; ord < statesByOrd.size(); ord++) {
+                    Object[] states = statesByOrd.get(ord);
+                    if (states == null) {
+                        continue;
+                    }
+                    BytesRef sharedKey = values.lookupOrd(ord);
+                    Object[] prevStates = statesByKey.get(sharedKey);
+                    if (prevStates == null) {
+                        long hashMapEntryOverhead = 36L;
+                        ramAccounting.addBytes(roundUp(
+                            BytesRefSizeEstimator.INSTANCE.estimateSize(sharedKey) + hashMapEntryOverhead));
+                        statesByKey.put(BytesRef.deepCopyOf(sharedKey), states);
+                    } else {
+                        for (int i = 0; i < aggregations.size(); i++) {
+                            AggregationContext aggregation = aggregations.get(i);
+                            //noinspection unchecked
+                            prevStates[i] = aggregation.function().reduce(
+                                ramAccounting,
+                                prevStates[i],
+                                states[i]
+                            );
+                        }
+                    }
+                }
+            }
+        }
+
+        if (nullStates != null) {
+            statesByKey.put(null, nullStates);
+        }
+        return statesByKey;
+    }
+
+    private static boolean docDeleted(@Nullable Bits liveDocs, int doc) {
+        return liveDocs != null && !liveDocs.get(doc);
+    }
+
+    private static void aggregateValues(List<AggregationContext> aggregations,
+                                        RamAccountingContext ramAccounting,
+                                        Object[] states) {
+        for (int i = 0; i < aggregations.size(); i++) {
+            AggregationContext aggregation = aggregations.get(i);
+            //noinspection unchecked
+            states[i] = aggregation.function().iterate(
+                ramAccounting,
+                states[i],
+                aggregation.inputs()
+            );
+        }
+    }
+
+    private static Object[] initStates(BigArrays bigArrays,
```

Comment:
can we be more exact on what "states" we actually initiate? this is super generic and without browsing the method body is not exactly clear what's doing

Reply:
I'm not sure. Any suggestions? I think `initAggregationStates` doesn't make it more readable as there are no other states in the scope of this class. Maybe we need to refer to the states of the aggregations with a different name, but I think it's quite fitting and clear.